server/src/main/java/org/elasticsearch/immutablestate/service/ImmutableClusterStateController.java

```diff
@@ -0,0 +1,302 @@
+     *        cluster state cannot be correctly applied, IncompatibleVersionException if the content is stale or
+     *        incompatible with this node {@link Version}, null if successful.
+     */
+    public void process(String namespace, XContentParser parser, Consumer<Exception> errorListener) {
+        Package immutableStatePackage;
+
+        try {
+            immutableStatePackage = packageParser.apply(parser, null);
+        } catch (Exception e) {
+            List<String> errors = List.of(e.getMessage());
+            recordErrorState(new ImmutableUpdateErrorState(namespace, -1L, errors, ImmutableStateErrorMetadata.ErrorKind.PARSING));
+            logger.error("Error processing state change request for [{}] with the following errors [{}]", namespace, errors);
+
+            errorListener.accept(new IllegalStateException("Error processing state change request for " + namespace, e));
+            return;
+        }
+
+        process(namespace, immutableStatePackage, errorListener);
+    }
+
+    /**
+     * Saves an immutable cluster state for a given 'namespace' from {@link Package}
+     *
+     * @param namespace the namespace under which we'll store the immutable keys in the cluster state metadata
+     * @param immutableStateFilePackage a {@link Package} composite state object to process
+     * @param errorListener a consumer called with IllegalStateException if the content has errors and the
+     *        cluster state cannot be correctly applied, IncompatibleVersionException if the content is stale or
+     *        incompatible with this node {@link Version}, null if successful.
+     */
+    public void process(String namespace, Package immutableStateFilePackage, Consumer<Exception> errorListener) {
+        Map<String, Object> immutableState = immutableStateFilePackage.state;
+        PackageVersion packageVersion = immutableStateFilePackage.metadata;
+
+        LinkedHashSet<String> orderedHandlers;
+        try {
+            orderedHandlers = orderedStateHandlers(immutableState.keySet());
+        } catch (Exception e) {
+            List<String> errors = List.of(e.getMessage());
+            recordErrorState(
+                new ImmutableUpdateErrorState(namespace, packageVersion.version(), errors, ImmutableStateErrorMetadata.ErrorKind.PARSING)
+            );
+            logger.error("Error processing state change request for [{}] with the following errors [{}]", namespace, errors);
+
+            errorListener.accept(new IllegalStateException("Error processing state change request for " + namespace, e));
+            return;
+        }
+
+        ClusterState state = clusterService.state();
+        ImmutableStateMetadata existingMetadata = state.metadata().immutableStateMetadata().get(namespace);
+        if (checkMetadataVersion(existingMetadata, packageVersion, errorListener) == false) {
+            return;
+        }
+
+        // Do we need to retry this, or it retries automatically?
```

Comment:
I have an outstanding question if there's anything else I need to do here to ensure the cluster state update is retried if it fails.

Reply:
Yes, there are no retries here, this state update will be attempted just once. Typically you would be calling this from a `TransportMasterNodeAction` which does handle certain retryable failures (i.e. `NotMasterException` and `FailedToCommitClusterStateException`) up to a timeout (defaulting to `30s`). This particular update can also fail in ways that don't make sense to retry. At the moment this doesn't seem to be called from any production code, retrying or otherwise.